src/Microsoft.Identity.Client/TokenCache.cs

```diff
@@ -204,1045 +207,1027 @@ internal void OnBeforeWrite(TokenCacheNotificationArgs args)
-                        TokenCacheAccessor.SaveRefreshToken(msalRefreshTokenCacheItem, requestParams.RequestContext);
-                    }
+                // if server returns the refresh token back, save it in the cache.
+                if (response.RefreshToken != null)
+                {
+                    msalRefreshTokenCacheItem = new MsalRefreshTokenCacheItem(preferredEnvironmentHost, requestParams.ClientId, response);
+                    requestParams.RequestContext.Logger.Info("Saving RT in cache...");
+                    TokenCacheAccessor.SaveRefreshToken(msalRefreshTokenCacheItem, requestParams.RequestContext);
+                }
 
-                    // save RT in ADAL cache for public clients
-                    // do not save RT in ADAL cache for MSAL B2C scenarios
-                    if (!requestParams.IsClientCredentialRequest && !requestParams.Authority.AuthorityType.Equals(Instance.AuthorityType.B2C))
-                    {
-                        CacheFallbackOperations.WriteAdalRefreshToken
-                            (LegacyCachePersistence, msalRefreshTokenCacheItem, msalIdTokenCacheItem,
-                            Authority.UpdateHost(requestParams.TenantUpdatedCanonicalAuthority, preferredEnvironmentHost),
-                            msalIdTokenCacheItem.IdToken.ObjectId, response.Scope);
-                    }
+                // save RT in ADAL cache for public clients
+                // do not save RT in ADAL cache for MSAL B2C scenarios
+                if (!requestParams.IsClientCredentialRequest && !requestParams.Authority.AuthorityType.Equals(Config.AuthorityType.B2C))
+                {
+                    CacheFallbackOperations.WriteAdalRefreshToken
+                        (Logger, LegacyCachePersistence, msalRefreshTokenCacheItem, msalIdTokenCacheItem,
+                        Authority.CreateAuthorityUriWithHost(requestParams.TenantUpdatedCanonicalAuthority, preferredEnvironmentHost),
+                        msalIdTokenCacheItem.IdToken.ObjectId, response.Scope);
+                }
 
-                    OnAfterAccess(args);
+                OnAfterAccess(args);
 
-                    return Tuple.Create(msalAccessTokenCacheItem, msalIdTokenCacheItem);
-                }
+                return Tuple.Create(msalAccessTokenCacheItem, msalIdTokenCacheItem);
+            }
                 finally
-                {
+            {
 #pragma warning disable CS0618 // Type or member is obsolete
                     HasStateChanged = false;
 #pragma warning restore CS0618 // Type or member is obsolete
                 }
-            }
         }
+    }
 
-        private void DeleteAccessTokensWithIntersectingScopes(AuthenticationRequestParameters requestParams,
-           ISet<string> environmentAliases, string tenantId, SortedSet<string> scopeSet, string homeAccountId)
+        internal ICoreLogger Logger => ServiceBundle.DefaultLogger;
```

Comment:
as mentioned elsewhere, it's a bit of off that it's internal... it might be relevant with a comment. Also consider moving to the other class properties